### PR TITLE
MULE-8753: Give the Extension model the concept of vendor

### PR DESCRIPTION
**Pull request by estebanwasinger**
This PR requires to have previously merged the following pull requests: https://github.com/mulesoft/mule-extensions-annotations/pull/25 and https://github.com/mulesoft/mule-extensions-api/pull/35

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/ImmutableExtensionModel.java [hunk 1]
```diff
@@ -67,6 +71,7 @@ protected ImmutableExtensionModel(String name,
 
         checkArgument(!StringUtils.isBlank(version), "version cannot be blank");
         this.version = version;
+        this.vendor = vendor;
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/ImmutableExtensionModel.java]: validate this is not null

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/DefaultExtensionManager.java [hunk 2]
```diff
@@ -123,20 +124,23 @@ public void stop() throws MuleException
...
+
+        LOGGER.info("Registering extension {} (version {})", extensionName, extensionVersion);
 
-        if (extensionRegistry.containsExtension(extensionName))
+        if (extensionRegistry.containsExtension(extensionName, extensionVendor))
         {
             if (LOGGER.isDebugEnabled())
             {
                 LOGGER.debug("A extension of name '{}' (version {}) is already registered. Skipping...",
-                             extensionModel.getName(), extensionModel.getVersion());
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/DefaultExtensionManager.java]: should add the vendor to this message as well?

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionEntityKey.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.internal.manager;
+
+import org.mule.extension.annotation.api.Extension;
+
+final class ExtensionEntityKey
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionEntityKey.java]: add javadoc

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionEntityKey.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+{
+
+    private String name;
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionEntityKey.java]: key fields should be final

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionEntityKey.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            return false;
+        }
+        ExtensionEntityKey entity = (ExtensionEntityKey) o;
+        return name.equals(entity.name) && vendor.equals(entity.vendor);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return name.hashCode() + vendor.hashCode();
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionEntityKey.java]: not a good hashcode function. Use multiplication instead and add a prime number into the mix. Or use the HashCodeBuilder from apache commons

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionEntityKey.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            vendor = Extension.NO_VENDOR;
+        }
+        this.name = name;
+        this.vendor = vendor;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (!(o instanceof ExtensionEntityKey))
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionEntityKey.java]: this is more clear to read if you remove the ! and the extra parenthesis and move the rest of the method inside the if block.

If the conditions doesn't checks out, you return false outside

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionRegistry.java [hunk 7]
```diff
@@ -151,7 +154,7 @@ void registerConfigurationProvider(ConfigurationProvider configurationProvider)
         catch (RegistrationException e)
         {
             throw new MuleRuntimeException(createStaticMessage(format(
-                    "Found exception while registering configuration provider '%'", configurationProvider.getName()))
+                    "Found exception while registering configuration provider '%s'", configurationProvider.getName()))
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/ExtensionRegistry.java]: cool catch! :)

File: modules/extensions-support/src/test/java/org/mule/module/extension/internal/manager/DefaultExtensionManagerTestCase.java [hunk 8]
```diff
@@ -61,6 +61,8 @@
 public class DefaultExtensionManagerTestCase extends AbstractMuleTestCase
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/module/extension/internal/manager/DefaultExtensionManagerTestCase.java]: add tests that:

* Verify you can't have a null vendor
* Extensions of same name but different vendor are not consider equal
* Exensions of same name and vendor are consider equal

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/ImmutableExtensionModel.java (hunk 1 above)
Comment (estebanwasinger) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/ImmutableExtensionModel.java]: Validating this property it's not entire clear, meanwhile using annotations "you can avoid the vendor configuration", when you use the fluent api, forcing the configuration of the vendor it's not intuitive to think that **No Vendor** means `.withVendor("")`, already exist the validation in `DefaultExtensionManager` which checks if it is null, and in this case, register the extension with a blank vendor.

Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/ImmutableExtensionModel.java]: I see what you mean but I don't see such validation on the DefaultExtensionManager. I think that if vendor cannot be null, then it should not be possible to instantiate this class on an invalid state, specially when the containsExtension() method depends on it. I think that the validation should be moved here

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/ImmutableExtensionModel.java [hunk 9]
```diff
@@ -65,8 +70,15 @@ protected ImmutableExtensionModel(String name,
         this.operations = toMap(operationModels);
         this.connectionProviders = immutableList(connectionProviders);
 
-        checkArgument(!StringUtils.isBlank(version), "version cannot be blank");
+        checkArgument(!StringUtils.isBlank(version), "Version cannot be blank");
+
+        if (vendor == null)
+        {
+            throw new IllegalModelDefinitionException("Vendor extension cannot be null");
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/ImmutableExtensionModel.java]: Rephrase as Vendor cannot be null

**Pull request merged**